### ISSUE TITLE
Dead keys do not work

### ISSUE THREAD

**Issue by samhed** [novnc/noVNC#350]
No dead keys work in Firefox on Linux. By dead keys I mean ~"^`' 
Tested on CentOS 6 and Fedora 18 with both Firefox 26 and 27.

It does work in other browsers such as Google Chrome and there is no problem on Windows (tested with Win8.1).

**Comment by hex-m** [comment]
As a workaround, you can disable dead keys on your client. Doesn't help in a multi-user environment but works for me. (Firefox on Xenial and noVNC on Proxmox 4.1)

**Comment by CendioOssman** [comment]
Unfortunately no, because:

 a) The browser might not be able to provide physical keys in all cases
 b) The server might not support this extension
 c) Sending physical keys is not a silver bullet, but rather a different trade-off. As such some users may prefer the symbol based approach.

**Comment by danielhb** [comment]
I don't think (b) is a problem nowadays - recent VNC servers have been supporting this extension for 5+ years, giving the original QEMU VNC extension was developed in 2010 or so.

But I agree with (a) and (c). I think this bug will have to wait for the solution of the Firefox bug mentioned a few comments above.

**Comment by CendioOssman** [comment]
I don't believe any of the RealVNC derived servers supports this, i.e. TigerVNC, TightVNC, TurboVNC, or the new proprietary RealVNC. So I'd say b) is very much an issue. Is there any server besides QEMU and libvncserver that supports it?

**Comment by danielhb** [comment]
You're right. Disregard my comment about (b).

**Comment by laggarcia** [comment]
> I don't believe any of the RealVNC derived servers supports this, i.e.
> TigerVNC, TightVNC, TurboVNC, or the new proprietary RealVNC. So I'd
> say b) is very much an issue. Is there any server besides QEMU and
> libvncserver that supports it?
>
>
Servers based on gtk-vnc library supports this extension.

**Comment by CendioOssman** [comment]
This is broken in Chrome as well, and likely in all browsers:

https://bugs.chromium.org/p/chromium/issues/detail?id=1135952
https://bugs.chromium.org/p/chromium/issues/detail?id=586568

**Comment by CendioOssman** [comment]
Also reported to the standard folks:

https://github.com/whatwg/dom/issues/900

**Comment by sergiomb2** [comment]
> Doesn't #21 fix this bug too?

I think  it not , I can't write an é (&eacute;)  and all others  accented keys in a novnc session  novnc-1.3.0

**Comment by sergiomb2** [comment]
but for me the question is ao I can type one accentuated characters ? , I found one workaround which is type the  accentuated character in clipboard of novnc .  Question can we map characters with on combination of keys ? characters like é ã ó ?

**Comment by sergiomb2** [comment]
I found  for alt-gr + 7 , 8 ,9 or 0  is another problem , is just when I use x11vnc instead vnc-server 

and repeting the question how you , type characters that need one dead key ?

**Comment by sergiomb2** [comment]
have we anyway to have composed keys like  ~ + a ? 

BTW I found x11vnc with -nomodtweak fix the altgr problem mention in my previous comment

**Comment by sergiomb2** [comment]
Hi, many thanks , on solution 1, is to apply on server ?

**Comment by kwibus** [comment]
@sergiomb2 the code fore solution 1 it bit of prove of concept.
* It makes deadkeys used in us-international work in linux, and probably mac (not yet tested),
* other keyboard layout support i still have to look into. if you have preferene for specicif layout, i could try that  as example for futher improvements
* I noticed it print some dead keys twice on windows, I think those might be  easy to fix

**Comment by sergiomb2** [comment]
With Firefox I got the console logs of the events and I managed to put all my dead keys working perfectly  .
but in Google chrome or electon don't work so well

**Comment by kwibus** [comment]
@sergiomb2 might be better to continue this conversation here: https://github.com/novnc/noVNC/pull/1907

**Comment by CendioOssman** [comment]
> I see 3 Possible Solution directions.

There is also the fourth option, using the QEMU raw keyboard event extension. It is supported by various servers.

It would move the keyboard layout handling to the server instead of the client. And dead keys work fine under that model, as far as I know.

**Comment by kwibus** [comment]
> There is also the fourth option, using the QEMU raw keyboard event extension. It is supported by various servers.

True, if I understand the code correctly, this will send `keysym`, and `scancode` instead of scancode.   
`keysym` is still unknown. 

I did my testing mostly with TigerVNC, there it tried to use QEMU extension, but it did not work. 
I assumed this was because `keysym` was 0, 
But i had not really checked if `Scancode` was set wrong.

I have to look into that, It might help.
